README.md
# filesystem
filesystem for golang

# installation

```
go get github.com/go-component/filesystem
```

# import
```
import "github.com/go-component/filesystem"
```


# Usage

support single or multiple for file operation

```

// single
Mkdir("a", 0755)

// multiple

Mkdir([]string{"a", "b"}, 0755)


```

# More Usage

#### func  AppendToFile

```go
func AppendToFile(fileName string, content []byte) error
```
Appends content to a file.

#### func  Chmod

```go
func Chmod(files interface{}, mode os.FileMode) error
```
Change mode for an array of files or directories.

#### func  ChmodWithRecur

```go
func ChmodWithRecur(files interface{}, mode os.FileMode) error
```
Change mode for an array of files or directories with recursive mode.

#### func  Chown

```go
func Chown(files interface{}, user, group int) error
```
Change the owner of an array of files or directories.

#### func  ChownWithRecur

```go
func ChownWithRecur(files interface{}, user, group int) error
```
Change the owner of an array of files or directories recursive mode.

#### func  Copy

```go
func Copy(srcFileName string, dstFileName string) error
```
Copies a file.

#### func  Dirname

```go
func Dirname(fileName string) string
```
Return dirname.

#### func  Exists

```go
func Exists(paths interface{}) bool
```
Checks the existence of files or directories.

#### func  Hardlink

```go
func Hardlink(srcFileName string, dstFileName string) error
```
Creates a hard link, or several hard links to files.

#### func  IsAbsolutePath

```go
func IsAbsolutePath(fileName string) bool
```
Return whether the file path is an absolute path.

#### func  IsDir

```go
func IsDir(path string) bool
```
Returns whether the file path is a directory.

#### func  IsFile

```go
func IsFile(path string) bool
```
Returns whether the file path is a file.

#### func  IsReadable

```go
func IsReadable(fileName string) bool
```
Tells whether a file exists and is readable.

#### func  IsWritable

```go
func IsWritable(fileName string) bool
```
Tells whether a file exists and is writable.

#### func  Mkdir

```go
func Mkdir(paths interface{}, mode os.FileMode) error
```
Creates a directory of directory or directories.

#### func  Readlink

```go
func Readlink(path string) (string, error)
```
Resolves links in paths.

#### func  Remove

```go
func Remove(files interface{}) error
```
Remove removes the named file or (empty) directory.

#### func  RemoveWithRecur

```go
func RemoveWithRecur(files interface{}) error
```
Remove removes the named file or directory with recursive mode.

#### func  Rename

```go
func Rename(srcFileName string, dstFileName string) error
```
Rename src to dst.

#### func  ResolveFilesAndDirs

```go
func ResolveFilesAndDirs(dirPath string) (files []string, dirs []string, err error)
```
Resolves files and directories.

#### func  Symlink

```go
func Symlink(srcDirName string, dstDirName string) error
```
Creates a symbolic link or copy a directory.

#### func  Touch

```go
func Touch(files interface{}) error
```
Creates new files if not exist.

#### func  TouchFromTime

```go
func TouchFromTime(files interface{}, atime time.Time, mtime time.Time) error
```
Sets access and modification time of files.


CHANGELOG.md
# 1.0.0
initialize project

filesystem_test.go
package filesystem

import (
	"os"
	"reflect"
	"testing"
	"time"
)

var dirRoot = "filesystem/test-mkdir"

var dirName = dirRoot + "/0"
var dirs = []string{
	dirRoot + "/1",
	dirRoot + "/2",
	dirRoot + "/3",
}

var recursiveDirRoot = dirRoot + "/sub-mkdir"

var recursiveDirName = recursiveDirRoot + "/0"

var recursiveDirs = []string{
	recursiveDirRoot + "/1",
	recursiveDirRoot + "/2",
	recursiveDirRoot + "/3",
}

func TestMkdir(t *testing.T) {

	err := Mkdir(dirName, 0755)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = Remove(dirs)
	})

	if !Exists(dirName) {
		t.Error("Mkdir test failed!")
	}

}

func TestMultiMkdir(t *testing.T) {

	err := Mkdir(dirs, 0755)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = Remove(dirs)
	})

	if !Exists(dirs) {
		t.Error("Multi Mkdir test failed!")
	}

}

func TestRemove(t *testing.T) {

	err := Mkdir(dirName, 0755)
	if err != nil {
		panic(err)
	}
	if !Exists(dirName) {
		t.Error("Remove test failed!")
	}
	err = Remove(dirName)
	if err != nil {
		panic(err)
	}

	if Exists(dirName) {
		t.Error("Remove test failed!")
	}
}

func TestMultiRemove(t *testing.T) {

	err := Mkdir(dirs, 0755)
	if err != nil {
		panic(err)
	}
	if !Exists(dirs) {
		t.Error("Multi Remove test failed!")
	}
	err = Remove(dirs)
	if err != nil {
		panic(err)
	}

	if Exists(dirs) {
		t.Error("Multi Remove test failed!")
	}
}

func TestRemoveWithRecur(t *testing.T) {

	err := Mkdir(recursiveDirName, 0755)
	if err != nil {
		panic(err)
	}
	if !Exists(recursiveDirName) {
		t.Error("Multi Remove test failed!")
	}
	err = RemoveWithRecur(recursiveDirRoot)
	if err != nil {
		panic(err)
	}

	if Exists(recursiveDirName) {
		t.Error("Multi Remove test failed!")
	}
}

func TestMultiRemoveWithRecur(t *testing.T) {

	err := Mkdir(recursiveDirs, 0755)
	if err != nil {
		panic(err)
	}

	if !Exists(recursiveDirs) {
		t.Error("Multi RemoveWithRecur test failed!")
	}
	err = RemoveWithRecur(recursiveDirRoot)
	if err != nil {
		panic(err)
	}

	if Exists(recursiveDirs) {
		t.Error("Multi RemoveWithRecur test failed!")
	}
}

func TestAppendToFile(t *testing.T) {
	fileName := dirRoot + "/append-to-file.txt"

	err := AppendToFile(fileName, []byte("filesystem"))
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(fileName) {
		t.Error("AppendToFile test failed!")
	}
}

func TestCopy(t *testing.T) {
	srcFileName, dstFileName := dirRoot+"/src-file.txt", dirRoot+"/dst-file.txt"

	err := AppendToFile(srcFileName, []byte("filesystem"))
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(srcFileName) {
		t.Error("Copy test failed!")
	}

	err = Copy(srcFileName, dstFileName)
	if err != nil {
		panic(err)
	}

	if !Exists(dstFileName) {
		t.Error("Copy test failed!")
	}

}

func TestTouch(t *testing.T) {
	fileName := dirRoot + "/touch.txt"

	err := Touch(fileName)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(fileName) {
		t.Error("Touch test failed!")
	}

}

func TestMultiTouch(t *testing.T) {
	files := []string{
		dirRoot + "/touch1.txt",
		dirRoot + "/touch2.txt",
		dirRoot + "/touch3.txt",
	}

	err := Touch(files)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(files) {
		t.Error("Touch test failed!")
	}
}

func TestTouchFromTime(t *testing.T) {
	fileName := dirRoot + "/touch.txt"

	now := time.Now().Local()

	err := TouchFromTime(fileName, now, now)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(fileName) {
		t.Error("TouchFromTime test failed!")
	}

	err = RemoveWithRecur(dirRoot)
	if err != nil {
		panic(err)
	}

	err = Touch(fileName)
	if err != nil {
		panic(err)
	}

	err = TouchFromTime(fileName, now, now)
	if err != nil {
		panic(err)
	}
}

func TestMultiTouchFromTime(t *testing.T) {
	files := []string{
		dirRoot + "/touch1.txt",
		dirRoot + "/touch2.txt",
		dirRoot + "/touch3.txt",
	}

	now := time.Now().Local()

	err := TouchFromTime(files, now, now)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(files) {
		t.Error("Multi TouchFromTime test failed!")
	}

	err = RemoveWithRecur(dirRoot)
	if err != nil {
		panic(err)
	}

	err = Touch(files)
	if err != nil {
		panic(err)
	}

	err = TouchFromTime(files, now, now)
	if err != nil {
		panic(err)
	}
}

func TestChmod(t *testing.T) {
	fileName := dirRoot + "/chmod.txt"

	err := Touch(fileName)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(fileName) {
		t.Error("Chmod test failed!")
	}

	err = Chmod(fileName, 0755)
	if err != nil {
		panic(err)
	}
}

func TestMultiChmod(t *testing.T) {
	files := []string{
		dirRoot + "/chmod1.txt",
		dirRoot + "/chmod2.txt",
		dirRoot + "/chmod3.txt",
	}

	err := Touch(files)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(files) {
		t.Error("Multi Chmod test failed!")
	}

	err = Chmod(files, 755)
	if err != nil {
		panic(err)
	}
}

func TestChmodWithRecur(t *testing.T) {
	fileName := recursiveDirRoot + "/chmod.txt"

	err := Touch(fileName)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(fileName) {
		t.Error("ChmodWithRecur test failed!")
	}

	err = ChmodWithRecur(dirRoot, 0755)
	if err != nil {
		panic(err)
	}
}

func TestMultiChmodWithRecur(t *testing.T) {
	fileName := []string{
		recursiveDirRoot + "/chmod1.txt",
		recursiveDirRoot + "/chmod2.txt",
		recursiveDirRoot + "/chmod3.txt",
	}

	err := Touch(fileName)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(fileName) {
		t.Error("Multi MultiChmodWithRecur test failed!")
	}

	err = ChmodWithRecur(dirRoot, 0755)
	if err != nil {
		panic(err)
	}
}

func TestChown(t *testing.T) {
	fileName := dirRoot + "/chown.txt"

	err := Touch(fileName)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(fileName) {
		t.Error("Chown test failed!")
	}

	err = Chown(fileName, 501, 20)
	if err != nil {
		panic(err)
	}
}

func TestMultiChown(t *testing.T) {
	files := []string{
		dirRoot + "/chown1.txt",
		dirRoot + "/chown2.txt",
		dirRoot + "/chown3.txt",
	}

	err := Touch(files)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(files) {
		t.Error("Multi Chown test failed!")
	}

	err = Chown(files, 501, 20)
	if err != nil {
		panic(err)
	}
}

func TestChownWithRecur(t *testing.T) {
	fileName := recursiveDirRoot + "/chown.txt"

	err := Touch(fileName)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(fileName) {
		t.Error("ChownWithRecur test failed!")
	}

	err = ChownWithRecur(dirRoot, 501, 20)
	if err != nil {
		panic(err)
	}
}

func TestMultiChownWithRecur(t *testing.T) {
	fileName := []string{
		recursiveDirRoot + "/chown1.txt",
		recursiveDirRoot + "/chown2.txt",
		recursiveDirRoot + "/chown3.txt",
	}

	err := Touch(fileName)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(fileName) {
		t.Error("Multi ChownWithRecur test failed!")
	}

	err = ChownWithRecur(dirRoot, 501, 20)
	if err != nil {
		panic(err)
	}
}

func TestRename(t *testing.T) {
	srcFileName, dstFileName := dirRoot+"/src-file.txt", dirRoot+"/dst-file.txt"

	err := AppendToFile(srcFileName, []byte("filesystem"))
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(srcFileName) {
		t.Error("Rename test failed!")
	}

	err = Rename(srcFileName, dstFileName)
	if err != nil {
		panic(err)
	}

	if !Exists(dstFileName) {
		t.Error("Rename test failed!")
	}
}

func TestIsDir(t *testing.T) {
	err := Mkdir(dirRoot, 0755)
	if err != nil {
		panic(err)
	}
	fileName := dirRoot + "/test.txt"
	err = Touch(fileName)
	if err != nil {
		panic(err)
	}
	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(dirRoot) {
		t.Error("IsDir test failed!")
	}

	if IsDir(dirRoot) != true {
		t.Error("IsDir test failed!")
	}
	if IsDir(fileName) == true {
		t.Error("IsDir test failed!")
	}
}

func TestIsFile(t *testing.T) {
	err := Mkdir(dirRoot, 0755)
	if err != nil {
		panic(err)
	}
	fileName := dirRoot + "/test.txt"
	err = Touch(fileName)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if !Exists(dirRoot) {
		t.Error("IsFile test failed!")
	}

	if IsFile(dirRoot) == true {
		t.Error("IsFile test failed!")
	}
	if IsFile(fileName) != true {
		t.Error("IsFile test failed!")
	}

}

func TestIsReadable(t *testing.T) {
	writeDir := "privilege" + "/read"
	err := Mkdir(writeDir, 0755)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur("privilege")
	})

	if !IsReadable(writeDir) {
		t.Error("IsReadable test failed!")
	}
}

func TestIsWriteable(t *testing.T) {

	writeDir := "privilege" + "/write"
	err := Mkdir(writeDir, 0755)
	if err != nil {
		panic(err)
	}

	t.Cleanup(func() {
		_ = RemoveWithRecur("privilege")
	})

	if !IsWritable(writeDir) {
		t.Error("IsWriteable test failed!")
	}
}


func TestHardlink(t *testing.T) {
	srcFileName := dirRoot + "/src.txt"
	dstFileName := dirRoot + "/dst.txt"
	err := Touch(srcFileName)
	if err != nil{
		panic(err)
	}

	err = Hardlink(srcFileName, dstFileName)
	if err != nil{
		panic(err)
	}

	srcFileInfo, err := os.Stat(srcFileName)

	if err != nil{
		panic(err)
	}

	dstFileInfo, err := os.Stat(dstFileName)

	if err != nil{
		panic(err)
	}


	srcNode := reflect.ValueOf(srcFileInfo.Sys()).Elem().Field(3).Uint()
	dstNode := reflect.ValueOf(dstFileInfo.Sys()).Elem().Field(3).Uint()


	t.Cleanup(func() {
		_ = RemoveWithRecur(dirRoot)
	})

	if srcNode != dstNode{
		t.Error("Hardlink test failed!")
	}
}


Filesystem.go
package filesystem

import (
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"path/filepath"
	"syscall"
	"time"
)

// creates a directory of directory or directories.
func Mkdir(paths interface{}, mode os.FileMode) {

	for _, pathName := range toIterable(paths) {
		err := os.MkdirAll(pathName, mode)
		if err != nil {
			panic(err)
		}
	}

}

// copies a file
func Copy(srcFileName string, dstFileName string) {

	srcFile, err := os.Open(srcFileName)
	if err != nil{
		panic(err)
	}

	dstFile, err := os.OpenFile(dstFileName, os.O_CREATE|os.O_WRONLY, 0666)

	if err != nil{
		panic(err)
	}

	_, err = io.Copy(dstFile, srcFile)

	if err != nil{
		panic(err)
	}
}

// checks the existence of files or directories
func Exists(paths interface{}) bool {
	for _, pathName := range toIterable(paths) {
		_, err := os.Stat(pathName) // For read access.
		if err != nil {
			return false
		}
	}

	return true
}

// creates new files
func Touch(files interface{}) {
	for _, fileName := range toIterable(files) {
		if !Exists(fileName) {
			file, err := os.Create(fileName)
			if err != nil {
				panic(file)
			}
			err = file.Close()
			if err != nil {
				panic(file)
			}
		}
	}
}

// sets access and modification time of files
func TouchFromTime(files interface{}, atime time.Time, mtime time.Time) {
	for _, fileName := range toIterable(files) {

		if Exists(fileName) {
			err := os.Chtimes(fileName, atime, mtime)
			if err != nil {
				panic(err)
			}
		} else {
			file, err := os.Create(fileName)
			if err != nil {
				panic(file)
			}
			err = file.Close()
			if err != nil {
				panic(file)
			}
		}
	}
}

// change mode for an array of files or directories.
func Chmod(files interface{}, mode os.FileMode) {
	for _, fileName := range toIterable(files) {

		if !Exists(fileName) {
			continue
		}
		err := os.Chmod(fileName, mode)
		if err != nil {
			panic(err)
		}
	}
}

// change mode for an array of files or directories with recursive
func ChmodWithRecur(files interface{}, mode os.FileMode) {

	for _, fileName := range toIterable(files) {

		if !Exists(fileName) {
			continue
		}
		if IsDir(fileName) {

			fs, ds := GetFilesAndDirs(fileName)

			ChmodWithRecur(append(fs, ds...), mode)

		} else {
			err := os.Chmod(fileName, mode)

			if err != nil {
				panic(err)
			}
		}
	}
}

// change the owner of an array of files or directories
func Chown(files interface{}, user, group int) {

	for _, fileName := range toIterable(files) {

		if !Exists(fileName) {
			continue
		}
		err := os.Chown(fileName, user, group)

		if err != nil {
			panic(err)
		}
	}
}

// change the owner of an array of files or directories recursive
func ChownWithRecur(files interface{}, user, group int) {

	for _, fileName := range toIterable(files) {

		if !Exists(fileName) {
			continue
		}

		if IsDir(fileName) {

			fs, ds := GetFilesAndDirs(fileName)

			ChownWithRecur(append(fs, ds...), user, group)

		} else {
			err := os.Chown(fileName, user, group)

			if err != nil {
				panic(err)
			}
		}
	}
}

// resolves files and directories
func GetFilesAndDirs(dirPath string) (files []string, dirs []string) {
	dir, err := ioutil.ReadDir(dirPath)
	if err != nil {
		panic(err)
	}

	PthSep := string(os.PathSeparator)

	for _, fi := range dir {
		if fi.IsDir() { // 目录, 递归遍历
			dirs = append(dirs, dirPath+PthSep+fi.Name())
			_, _ = GetFilesAndDirs(dirPath + PthSep + fi.Name())

		} else {
			files = append(files, dirPath+PthSep+fi.Name())
		}
	}

	return files, dirs
}

// rename src to dst
func Rename(srcFileName string, dstFileName string) {
	err := os.Rename(srcFileName, dstFileName)

	if err != nil{
		panic(err)
	}
}

// returns whether the file path is a directory
func IsDir(path string) bool {
	s, err := os.Stat(path)
	if err != nil {
		return false
	}
	return s.IsDir()
}

// returns whether the file path is a file
func IsFile(path string) bool {
	return !IsDir(path)
}

// tells whether a file exists and is readable
func IsReadable(fileName string) bool {
	err := syscall.Access(fileName, syscall.O_RDONLY)
	if err != nil {
		return false
	}

	return true
}

// tells whether a file exists and is writable
func IsWritable(fileName string) bool {

	err := syscall.Access(fileName, syscall.O_RDWR)
	if err != nil {
		return false
	}

	return true
}

// creates a symbolic link or copy a directory
func Symlink(srcDirName string, dstDirName string) {
	err := os.Symlink(srcDirName, dstDirName)

	if err != nil{
		panic(err)
	}
}

// creates a hard link, or several hard links to a file
func Hardlink(srcFileName string, dstFileNames interface{}) {

}
// resolves links in paths.
func Readlink(path string) string{
	link,err := os.Readlink(path)

	if err != nil{
		panic(err)
	}
	return link
}

// returns whether the file path is an absolute path
func IsAbsolutePath(fileName string) bool {
	return filepath.IsAbs(fileName)
}

// get dirname
func Dirname(fileName string) string {
	return filepath.Dir(fileName)
}

// appends content to a file
func AppendToFile(fileName string, content []byte) {

	dir := Dirname(fileName)

	if !Exists(dir) {
		Mkdir(dir, 0666)
	}

	if !IsWritable(fileName) {
		panic(fmt.Sprintf("Unable to write to the \"%s\" directory.", fileName))
	}

	file, err := os.OpenFile(fileName, os.O_APPEND|os.O_CREATE, 0666)

	if err != nil{
		panic(err)
	}

	_, err = file.Write(content)

	if err != nil {
		panic(err)
	}
}

func toIterable(files interface{}) []string {
	switch files.(type) {
	case string:
		return []string{files.(string)}

	case []string:
		return files.([]string)
	default:
		panic("invalid interface type")
	}
}


filesystem.go
package filesystem

import (
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"path/filepath"
	"syscall"
	"time"
)

// Creates a directory of directory or directories.
func Mkdir(paths interface{}, mode os.FileMode) error {

	for _, pathName := range toIterable(paths) {

		if Exists(pathName) {
			continue
		}

		err := os.MkdirAll(pathName, mode)
		if err != nil {
			return err
		}
	}

	return nil
}

// Remove removes the named file or (empty) directory.
func Remove(files interface{}) error {

	for _, fileName := range toIterable(files) {

		if fileName == "" {
			continue
		}

		if !Exists(fileName) {
			continue
		}

		err := os.Remove(fileName)

		if err != nil {
			return err
		}
	}
	return nil
}

// Remove removes the named file or directory with recursive mode.
func RemoveWithRecur(files interface{}) error {

	for _, fileName := range toIterable(files) {

		if fileName == "" {
			continue
		}

		if !Exists(fileName) {
			continue
		}

		if IsDir(fileName) {
			fs, ds, err := ResolveFilesAndDirs(fileName)
			if err != nil {
				return err
			}

			moreFiles := append(fs, ds...)
			if len(moreFiles) > 0 {
				err := RemoveWithRecur(moreFiles)
				if err != nil {
					return err
				}
			}
		}

		err := os.Remove(fileName)

		if err != nil {
			return err
		}
	}

	return nil
}

// Copies a file.
func Copy(srcFileName string, dstFileName string) error {

	srcFile, err := os.Open(srcFileName)
	if err != nil {
		return err
	}

	dstFile, err := os.OpenFile(dstFileName, os.O_CREATE|os.O_WRONLY, 0666)

	if err != nil {
		return err
	}

	_, err = io.Copy(dstFile, srcFile)

	if err != nil {
		return err
	}

	return nil
}

// Checks the existence of files or directories.
func Exists(paths interface{}) bool {

	for _, pathName := range toIterable(paths) {
		_, err := os.Stat(pathName) // For read access.
		if err != nil {
			return false
		}
	}

	return true
}

// Creates new files if not exist.
func Touch(files interface{}) error {

	for _, fileName := range toIterable(files) {

		dir := Dirname(fileName)
		if !Exists(dir) {
			err := Mkdir(dir, 0755)
			if err != nil {
				return err
			}
		}

		if !Exists(fileName) {
			file, err := os.Create(fileName)
			if err != nil {
				return err
			}
			err = file.Close()
			if err != nil {
				return err
			}
		}
	}

	return nil
}

// Sets access and modification time of files.
func TouchFromTime(files interface{}, atime time.Time, mtime time.Time) error {

	for _, fileName := range toIterable(files) {
		dir := Dirname(fileName)
		if !Exists(dir) {
			err := Mkdir(dir, 0755)
			if err != nil {
				return err
			}
		}
		if Exists(fileName) {
			err := os.Chtimes(fileName, atime, mtime)
			if err != nil {
				return err
			}
		} else {
			file, err := os.Create(fileName)
			if err != nil {
				return err
			}
			err = file.Close()
			if err != nil {
				return err
			}
			err = TouchFromTime(fileName, atime, mtime)
			if err != nil {
				return err
			}
		}
	}

	return nil
}

// Change mode for an array of files or directories.
func Chmod(files interface{}, mode os.FileMode) error {

	for _, fileName := range toIterable(files) {

		if !Exists(fileName) {
			continue
		}
		err := os.Chmod(fileName, mode)
		if err != nil {
			return err
		}
	}
	return nil
}

// Change mode for an array of files or directories with recursive mode.
func ChmodWithRecur(files interface{}, mode os.FileMode) error {

	for _, fileName := range toIterable(files) {

		if !Exists(fileName) {
			continue
		}
		if IsDir(fileName) {

			fs, ds, err := ResolveFilesAndDirs(fileName)
			if err != nil {
				return err
			}

			err = ChmodWithRecur(append(fs, ds...), mode)
			if err != nil {
				return err
			}

		} else {
			err := os.Chmod(fileName, mode)

			if err != nil {
				return err
			}
		}
	}

	return nil
}

// Change the owner of an array of files or directories.
func Chown(files interface{}, user, group int) error {

	for _, fileName := range toIterable(files) {

		if !Exists(fileName) {
			continue
		}
		err := os.Chown(fileName, user, group)

		if err != nil {
			return err
		}
	}

	return nil
}

// Change the owner of an array of files or directories recursive mode.
func ChownWithRecur(files interface{}, user, group int) error {

	for _, fileName := range toIterable(files) {

		if !Exists(fileName) {
			continue
		}

		if IsDir(fileName) {

			fs, ds, err := ResolveFilesAndDirs(fileName)
			if err != nil {
				return err
			}

			err = ChownWithRecur(append(fs, ds...), user, group)
			if err != nil {
				return err
			}

		} else {
			err := os.Chown(fileName, user, group)

			if err != nil {
				return err
			}
		}
	}

	return nil
}

// Resolves files and directories.
func ResolveFilesAndDirs(dirPath string) (files []string, dirs []string, err error) {

	dir, err := ioutil.ReadDir(dirPath)
	if err != nil {
		return []string{}, []string{}, err
	}

	PathSep := string(os.PathSeparator)

	for _, fi := range dir {
		if fi.IsDir() {
			dirs = append(dirs, dirPath+PathSep+fi.Name())
			_, _, _ = ResolveFilesAndDirs(dirPath + PathSep + fi.Name())

		} else {
			files = append(files, dirPath+PathSep+fi.Name())
		}
	}

	return files, dirs, nil
}

// Rename src to dst.
func Rename(srcFileName string, dstFileName string) error {

	err := os.Rename(srcFileName, dstFileName)

	if err != nil {
		return err
	}

	return nil
}

// Returns whether the file path is a directory.
func IsDir(path string) bool {

	s, err := os.Stat(path)
	if err != nil {
		return false
	}
	return s.IsDir()
}

// Returns whether the file path is a file.
func IsFile(path string) bool {

	return !IsDir(path)
}

// Tells whether a file exists and is readable.
func IsReadable(fileName string) bool {

	err := syscall.Access(fileName, syscall.O_RDONLY)
	if err != nil {
		return false
	}

	return true
}

// Tells whether a file exists and is writable.
func IsWritable(fileName string) bool {

	err := syscall.Access(fileName, syscall.O_RDWR)
	if err != nil {
		return false
	}

	return true
}

// Creates a symbolic link or copy a directory.
func Symlink(srcDirName string, dstDirName string) error {
	err := os.Symlink(srcDirName, dstDirName)

	if err != nil {
		return err
	}

	return nil
}

// Creates a hard link, or several hard links to files.
func Hardlink(srcFileName string, dstFileName string) error {
	err := os.Link(srcFileName, dstFileName)
	if err != nil {
		return err
	}
	return nil
}

// Resolves links in paths.
func Readlink(path string) (string, error) {
	link, err := os.Readlink(path)

	if err != nil {
		return "", err
	}
	return link, nil
}

// Return whether the file path is an absolute path.
func IsAbsolutePath(fileName string) bool {
	return filepath.IsAbs(fileName)
}

// Return dirname.
func Dirname(fileName string) string {
	return filepath.Dir(fileName)
}

// Appends content to a file.
func AppendToFile(fileName string, content []byte) error {

	dir := Dirname(fileName)
	if !Exists(dir) {
		err := Mkdir(dir, 0755)
		if err != nil {
			return err
		}
	}

	if !IsWritable(dir) {
		panic(fmt.Sprintf("Unable to write to the \"%s\" directory.", fileName))
	}

	file, err := os.OpenFile(fileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0666)
	if err != nil {
		return err
	}

	_, err = file.Write(content)

	if err != nil {
		return err
	}

	return nil
}

func toIterable(files interface{}) []string {
	switch files.(type) {
	case string:
		return []string{files.(string)}

	case []string:
		return files.([]string)
	default:
		panic("invalid interface type")
	}
}
